### DISCUSSION TITLE
Inlay Synthetics

### DISCUSSION THREAD

**Discussion by TimWSpence** [scalameta/nvim-metals#576] [General]
I've got synthetics (inferred type, implicits, etc) working in LSP hover but was wondering if there is a way to display them as inlay? (I _think_ that's the term. I've attached a picture of what I get with Rust which is the kind of thing I was thinking of)
<img width="886" alt="Screenshot 2023-06-12 at 18 54 24" src="https://github.com/scalameta/nvim-metals/assets/3360080/68c6658d-7f94-4dee-832a-b410efea4747">

**Comment by ckipp01** [comment]
Good question @TimWSpence. So Metals _does_ actually support this, but early on Neovim didn't unless you put the result at the end of the line. When playing around with this it resulted in some really confusing behavior at times because if you had some code like:

```scala
val result = List(1, 2, 3).map(_.toDouble).map(_.toString)
```

And you tried to show the inferred types you could end up with ` // : List[Int] : List[String]` jammed all on to the end of the line _instead_ of being inlined. The same with showing implicits when there are multiple in the same line. Because of this, for editors that didn't support inline virtual text, we opted to put it in comments.  Now, with that being said, now that https://github.com/neovim/neovim/pull/20130 is merged, this could probably totally change. nvim-metals _could_ try to display it in the same way that VS Code does, _or_ it could also be adapted to use `inlayHints`. There was a POC for this done by @tgodzik somewhere and @rochala has already actually played around with this in nvim-metals!

https://github.com/scalameta/nvim-metals/assets/13974112/01812731-bdda-41e4-9fc7-bbf1c0b06c76

So, right now it's not really supported, but in the future it totally could be.

**Reply by rochala** [reply]
Actually, I have great news @jkciesluk started porting Decoration Protocol to inlay hints, so we will have first-hand support in metals soon. AFAIK the support for virtual text in between of text is going to be shipped with next version of neovim. I'm not sure if there is going to be first hand LSP support for it or whether we will be forced to use some plugins.

Also, it is worth noting, that inlay hints are way more powerful than Decoration protocol as they allow hovering and go to definition on displayed types.
